### PR TITLE
Added Embed functionality

### PR DESCRIPTION
**Pull request by xiluisx**
Plz check my pull request Mr.G.

### CODE REVIEW

File: GHooks/settings.json [hunk 1]
```diff
@@ -1,40 +1,138 @@
...
-     
-     //please put the web hook link here, leave blank for disable
-     "GlobalChat" : "",
-     "GlobalFormat": "{0}",
 
-     //please put the web hook link here, leave blank for disable
-     "LocalChat": "",
-     "LocalFormat": "local: {0}",
+      //please put the web hook link here, leave blank for disable
+      "GlobalChat": "https://discordapp.com/api/webhooks/741815178844176416/2XL_m-BMsr6n2DpRaOHNanjgo-B8rG8gRrV_Yhxeko2R06dswsUTCscgbiOaHbmgyyrz",
```
Comment (ggggg) [inline GHooks/settings.json]:
```suggestion
      "GlobalChat": "",
```

File: GHooks/settings.json [hunk 2]
```diff
@@ -1,40 +1,138 @@
...
+         "image": null,
+         "thumbnail": null,
+         "video": null,
+         "provider": null,
+         "author": null,
+         "fields": []
+      }],
+
+      //please put the web hook link here, leave blank for disable
+      "LocalChat": "https://discordapp.com/api/webhooks/741815178844176416/2XL_m-BMsr6n2DpRaOHNanjgo-B8rG8gRrV_Yhxeko2R06dswsUTCscgbiOaHbmgyyrz",
```
Comment (ggggg) [inline GHooks/settings.json]:
```suggestion
      "LocalChat": "",
```

File: GHooks/settings.json [hunk 3]
```diff
@@ -1,40 +1,138 @@
...
+         "provider": null,
+         "author": null,
+         "fields": []
+      }],
 
-     //please put the web hook link here, leave blank for disable
-     "CommandsLog" : "",
-     "CommandsLogFormat" : "{0}"
+      //please put the web hook link here, leave blank for disable
+      "CommandsLog": "https://discordapp.com/api/webhooks/741815178844176416/2XL_m-BMsr6n2DpRaOHNanjgo-B8rG8gRrV_Yhxeko2R06dswsUTCscgbiOaHbmgyyrz",
```
Comment (ggggg) [inline GHooks/settings.json]:
```suggestion
      "CommandsLog": "",
```

File: src/Webhooks/Core.cs [hunk 4]
```diff
@@ -3,9 +3,11 @@
 using BrokeProtocol.API;
 using BrokeProtocol.Entities;
 using BrokeProtocol.Managers;
+using Newtonsoft.Json;
 using System;
 using System.Collections.Generic;
 using System.Diagnostics;
+using System.IO;//remove
```
Comment (ggggg) [inline src/Webhooks/Core.cs]: // remove?

File: src/Webhooks/Events/OnGlobalChatMessage.cs [hunk 5]
```diff
@@ -8,11 +9,24 @@ public class OnGlobalChatMessage : IScript
         [Target(GameSourceEvent.PlayerGlobalChatMessage, ExecutionMode.Event)]
         public void OnEvent(ShPlayer player, string message)
         {
+            List<Embed> commandsEmb = new List<Embed>();
+            List<Embed> globalEmb = new List<Embed>();
```
Comment (ggggg) [inline src/Webhooks/Events/OnGlobalChatMessage.cs]: Use var where ever possible 
```suggestion
           var commandsEmb = new List<Embed>();
           var globalEmb = new List<Embed>();
```

File: src/Webhooks/Events/OnLeave.cs [hunk 6]
```diff
@@ -8,7 +9,14 @@ public class OnLeave : IScript
...
         {
-            Core.Instance.joinWebhook.Send(string.Format(Core.Instance.Settings.Server.PlayerLeaveFormat, player.username), player.username);
+            List<Embed> leaveEmb = new List<Embed>();
+
+            foreach (var em in Core.Instance.Settings.Server.PlayerLeaveEmbed)
+            {
+                leaveEmb.Add(new Embed { Title = string.Format(em.Title, player.username), Description = string.Format(em.Description, player.username) });
+            }
+            Core.Instance.joinWebhook.Send(string.Format(Core.Instance.Settings.Server.PlayerLeaveFormat, player.username), player.username,
+                embeds: Core.Instance.Settings.Server.PlayerLeaveUseEmbed? leaveEmb:null);
```
Comment (ggggg) [inline src/Webhooks/Events/OnLeave.cs]: Seems like this is a lot of duplicate code, maybe we would have it in a function that takes in the embed settings and returns it back:
  ```cs
public List<Embed> CreateAllEmbeds(List<Embed> embeds) {
            var emb = new List<Embed>();
            foreach (var em in Core.Instance.Settings.Server.PlayerLeaveEmbed)
            {
                emb.Add(new Embed { Title = string.Format(em.Title, player.username), Description = string.Format(em.Description, player.username) }); // add more things to format and make docs on what each mean in the json
            }
}
 ```

File: src/Webhooks/Events/OnLeave.cs [hunk 7]
```diff
@@ -8,7 +9,14 @@ public class OnLeave : IScript
         [Target(GameSourceEvent.PlayerDestroy, ExecutionMode.Event)]
         public void OnEvent(ShPlayer player)
         {
-            Core.Instance.joinWebhook.Send(string.Format(Core.Instance.Settings.Server.PlayerLeaveFormat, player.username), player.username);
+            List<Embed> leaveEmb = new List<Embed>();
```
Comment (ggggg) [inline src/Webhooks/Events/OnLeave.cs]:
```suggestion
            var leaveEmb = new List<Embed>();
```

File: src/Webhooks/Events/OnLocalChatMessage.cs [hunk 8]
```diff
@@ -8,11 +9,23 @@ public class OnLocalChatMessage : IScript
...
+                localEmb.Add(new Embed { Title = string.Format(em.Title, message, player.username), Description = string.Format(em.Description, message, player.username) });
+            }
             if (message.StartsWith("/"))
             {
-                Core.Instance.commandWebhook.Send(string.Format(Core.Instance.Settings.Chat.CommandsLogFormat, message, player.username), player.username);
+                Core.Instance.commandWebhook.Send(string.Format(Core.Instance.Settings.Chat.CommandsLogFormat, message, player.username), player.username,embeds: Core.Instance.Settings.Chat.CommandsUseEmbed?commandsEmb:null);
+                return;
             }
-            Core.Instance.localWebhook.Send(string.Format(Core.Instance.Settings.Chat.LocalFormat, message, player.username), player.username);
+            Core.Instance.localWebhook.Send(string.Format(Core.Instance.Settings.Chat.LocalFormat, message, player.username), player.username,embeds: Core.Instance.Settings.Chat.LocalUseEmbed?localEmb:null);
```
Comment (ggggg) [inline src/Webhooks/Events/OnLocalChatMessage.cs]: Should be:
```cs
            if (message.StartsWith("/"))
            {
            foreach (var em in Core.Instance.Settings.Chat.CommandsEmbed)
            {
                commandsEmb.Add(new Embed { Title = string.Format(em.Title, message, player.username), Description = string.Format(em.Description, message, player.username) });
            }
Core.Instance.commandWebhook.Send(string.Format(Core.Instance.Settings.Chat.CommandsLogFormat, message, player.username), player.username,embeds: Core.Instance.Settings.Chat.CommandsUseEmbed?commandsEmb:null);
                return;
            }
            foreach (var em in Core.Instance.Settings.Chat.LocalEmbed)
            {
                localEmb.Add(new Embed { Title = string.Format(em.Title, message, player.username), Description = string.Format(em.Description, message, player.username) });
            }
            Core.Instance.localWebhook.Send(string.Format(Core.Instance.Settings.Chat.LocalFormat, message, player.username), player.username,embeds: Core.Instance.Settings.Chat.LocalUseEmbed?localEmb:null);
``` 
I guess

File: src/Webhooks/Events/OnLogin.cs [hunk 9]
```diff
@@ -10,7 +9,13 @@ public class OnLogin : IScript
         [Target(GameSourceEvent.PlayerInitialize, ExecutionMode.Event)]
         public void OnEvent(ShPlayer player)
         {
-            Core.Instance.joinWebhook.Send(string.Format(Core.Instance.Settings.Server.PlayerJoinFormat, player.username), player.username);
+            List<Embed> joinEmb = new List<Embed>();
```
Comment (ggggg) [inline src/Webhooks/Events/OnLogin.cs]:
```suggestion
            var joinEmb = new List<Embed>();
```

File: src/Webhooks/Properties/AssemblyInfo.cs [hunk 10]
```diff
@@ -12,4 +12,4 @@
 [assembly: ComVisible(false)]
 [assembly: Guid("6898544b-a280-4866-ba6f-9d0dc8f62ca8")]
 
-[assembly: AssemblyFileVersion("1.0.0.0")]
+[assembly: AssemblyFileVersion("2.0.0.0")]
```
Comment (ggggg) [inline src/Webhooks/Properties/AssemblyInfo.cs]:
```suggestion
[assembly: AssemblyFileVersion("1.2.0.0")]
```
idk

**Review by ggggg — CHANGES_REQUESTED** [review]

**Comment by ggggg** [comment]
You tested everything?
Because idk if the settings json can have embed as a type.

File: src/Webhooks/Core.cs (hunk 4 above)
Comment (xiluisx) [reply src/Webhooks/Core.cs]: Added when testing so no need to have it

File: src/Webhooks/Core.cs [hunk 11]
```diff
@@ -61,8 +61,10 @@ public void RegisterCustomCommands()
             foreach (var customEvent in CustomEventReader.Parsed)
             {
                 Logger.LogInfo($"[CC] Registering custom event(s) for webhook {string.Join(", ", customEvent.Event)} by name '{customEvent.Event}'..");
-                EventsHandler.Add(customEvent.Event, new Action<ShPlayer, string>((player, eventName) => {
-                    if (player.HasPermission("webhook." + eventName)){
+                EventsHandler.Add(customEvent.Event, new Action<ShPlayer, string>((player, eventName) =>
+                {
+                    if (player.svPlayer.HasPermission("webhook." + eventName))
+                    {
```
Comment (ggggg) [inline src/Webhooks/Core.cs:67]: We might wanna add embeds to that soon... not urgent though.

File: src/Webhooks/EmbedCrafter.cs [hunk 12]
```diff
@@ -0,0 +1,68 @@
+﻿using BrokeProtocol.Entities;
+using System.Collections.Generic;
+using System.Linq;
+
+namespace Webhooks
+{
+    public class EmbedCrafter
```
Comment (ggggg) [inline src/Webhooks/EmbedCrafter.cs]:
```suggestion
    public static class EmbedCrafter
```

**Review by ggggg — CHANGES_REQUESTED** [review]

**Review by ggggg — APPROVED** [review]

**Comment by xiluisx** [comment]
accept now?

**Pull request merged**